### PR TITLE
ROC-3261: Save payment declaration for defence in case claim amount is paid

### PR DESCRIPTION
**Pull request by rattur**
Initial server-side components and files for saving full defence with when did you pay

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ROC-3261

### Change description ###

Updated Full defence response model to include when did you pay

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/response/ResponseMapper.java [hunk 1]
```diff
@@ -71,7 +71,8 @@ public FullDefenceResponse from(CCDResponse response) {
             partyMapper.from(response.getDefendant()),
             statementOfTruth,
             FullDefenceResponse.DefenceType.valueOf(response.getResponseType().name()),
-            response.getDefence()
+            response.getDefence(),
+            null
```
Comment (geek-soft) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/response/ResponseMapper.java]: That is not right, you basically ignored new property in CCD world. Please implement storing the payment details in CCD as well.

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/FullDefenceResponse.java [hunk 2]
```diff
@@ -22,17 +23,22 @@
     @Size(max = 99000)
     private final String defence;
 
+    @Valid
+    private final WhenDidYouPay whenDidYouPay;
```
Comment (geek-soft) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/FullDefenceResponse.java]: Please rename it to be `PaymentDeclaration` or `PaymentDetails`.

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonUnwrapped;
+import org.hibernate.validator.constraints.NotBlank;
+import uk.gov.hmcts.cmc.domain.constraints.DateNotInTheFuture;
+import java.time.LocalDate;
+import java.util.Objects;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+
+    @JsonInclude(JsonInclude.Include.NON_ABSENT)
```
Comment (geek-soft) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java]: Please remove.

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    public class WhenDidYouPay {
+
+        @JsonUnwrapped
```
Comment (geek-soft) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java]: Please remove.

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        @NotNull
+        @DateNotInTheFuture
+        private final LocalDate paidDate;
+
+        @NotBlank
+        @Size(max = 99000)
+        private final String explanation;
+
+        public WhenDidYouPay(final LocalDate paidDate, String explanation) {
```
Comment (geek-soft) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java]: Remove final modifier.

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(paidDate, explanation);
+        }
+
+        @Override
+        public String toString() {
+            return String.format(
```
Comment (geek-soft) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPay.java]: Use `return ReflectionToStringBuilder.toString(this, ourStyle())`, we do not use `String.format`

File: domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            .withExplanation("defence")
+            .withPaidDate(LocalDate.now().plusYears(1))
+            .build();
+        //when
+        Set<String> errors = validate(whenDidYouPay);
+        //then
+        assertThat(errors)
+            .hasSize(1)
+            .contains("paidDate : is in the future");
+
```
Comment (geek-soft) [inline domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java]: Remove extra blank line.

File: domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
```
Comment (geek-soft) [inline domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java]: Remove extra blank line.

File: domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            .withExplanation("defence")
+            .withPaidDate(null)
+            .build();
+        //when
+        Set<String> errors = validate(whenDidYouPay);
+        //then
+        assertThat(errors)
+            .hasSize(1)
+            .contains("paidDate : may not be null");
+
```
Comment (geek-soft) [inline domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java]: Remove blank line.

File: domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    @Test
+    public void shouldHaveNoErrorsWhenThereIsValidDateAndExplanation() {
```
Comment (geek-soft) [inline domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/WhenDidYouPayTest.java]: That duplicates first test case in the test file, remove one of these.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/response/ResponseMapper.java (hunk 1 above)
Comment (rattur) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/response/ResponseMapper.java]: Will speak to Sabah about this as we decided to get data saved to claim store first

Comment (geek-soft) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/response/ResponseMapper.java]: Who decided it? At this point CCD data store gets the same level of support as database.

File: domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclarationTest.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        //when
+        Set<String> errors = validate(paymentDeclaration);
+        //then
+        assertThat(errors)
+            .hasSize(1)
+            .contains("explanation : may not be empty");
+    }
+
+    @Test
+    public void shouldHaveValidationMessagesWhenExplanationExceedsSizeLimint() {
```
Comment (geek-soft) [inline domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclarationTest.java]: Typo in last word.

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclaration.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+        @NotNull
+        @DateNotInTheFuture
+        private final LocalDate paidDate;
+
+        @NotBlank
+        @Size(max = 99000)
+        private final String explanation;
+
+        public PaymentDeclaration(final LocalDate paidDate, String explanation) {
```
Comment (geek-soft) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclaration.java]: Remove final modifier.

File: domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclarationTest.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        //given
+        PaymentDeclaration paymentDeclaration = SamplePaymentDeclaration.validDefaults();
+        //when
+        Set<String> response = validate(paymentDeclaration);
+        //then
+        assertThat(response).hasSize(0);
+    }
+
+    @Test
+    public void shouldBeInvalidForNullExplanation() {
```
Comment (geek-soft) [inline domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclarationTest.java]: Rename to `shouldHaveValidationMessageWhenExplanationIsNull` for consistency.

File: domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclarationTest.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+import java.time.LocalDate;
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.hmcts.cmc.domain.BeanValidator.validate;
+
+public class PaymentDeclarationTest {
+
+    @Test
+    public void shouldBeSuccessfulValidationForWhenDidYouPay() {
```
Comment (geek-soft) [inline domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/PaymentDeclarationTest.java]: Rename to `shouldHaveNoValidationMessageWhenInstanceIsValid`

File: domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleResponse.java [hunk 15]
```diff
@@ -10,6 +11,7 @@
     public static class FullDefence extends SampleResponse<FullDefence> {
         private FullDefenceResponse.DefenceType defenceType = FullDefenceResponse.DefenceType.DISPUTE;
         private String defence = "defence string";
+        private PaymentDeclaration paymentDeclaration = SamplePaymentDeclaration.builder().build();
```
Comment (geek-soft) [inline domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleResponse.java]: Please add method that allows overriding `paymentDeclaration` value e.g. `withPaymentDeclaration`

**Pull request merged**